### DISCUSSION TITLE
Change Dagster log level on runtime

### DISCUSSION THREAD

**Discussion by honno** [dagster-io/dagster#7974] [Q&A] [answered]
Hiya folks, I was wondering how one could change the log level on runtime. My specific context is running tests, where the default log level will result in a lot of noisy stdout on `pytest` failures.

```python
def test_smoke_pipeline():
    res = etl_pipeline.execute_in_process(...)
    assert res.success
```

My mental model is admittedly weak when it comes to contexts. I couldn't quite work out creating my own logger and setting the level then, as there's things like setting a handler that I don't understand. Happy to do more investigation on this, but wonder if there's a quick win here.

I want to keep the log level as its default when running the pipeline generally, so a `config.yml` doesn't seem the right solution. I'd rather not swap configs via a CLI flag or something.

cc @leej3

**Comment by jamiedemaria** [comment, marked as answer]
Hi there! the docs for setting the log level can be found here https://docs.dagster.io/concepts/logging/loggers#configuring-the-built-in-loggers

The example shows config in yaml format, but you can also provide this as a `run_config` dict to `execute_in_process`. It would look something like this

```
run_config = {"loggers": {"console": {"config": {"log_level": "ERROR"}}}}
etl_pipeline.execute_in_process(run_config=run_config)
```

**Reply by honno** [reply]
Oh awesome, thanks! I hadn't caught that `run_config` mirrored the `config.yml` semantics :facepalm:

**Comment by zyd14** [comment]
@jamiedemaria I tried using this suggestion by adding 
```
loggers:
  console:
    config:
      log_level: WARNING (I also tried ERROR)
```
to my run config, but noticed no changes after launching a run - DEBUG and INFO logs are still getting logged to the console UI. I'm using the normal Dagster logger retrieved from `context.log` and `get_dagster_logger`. Any ideas on why this isn't working? I'd like to avoid changing the log level globally as I only have one job that needs reduced logging. I've also tried just manually setting the level in the op like `context.log.setLevel("WARNING")`, which did not work.

**Comment by SbstnErhrdt** [comment]
I am also struggling with the logging.

The solution of @jamiedemaria does not work for me:


```python
import dagster as dg

@dg.job()
def job():
    logger = dg.get_dagster_logger()
    logger.info(f"INFO LOG MSG")
    logger.error(f"ERROR LOG MSG")
    logger.warning(f"WARNING LOG MSG")
    logger.debug(f"DEBUG LOG MSG")


if __name__ == "__main__":
    run_config = {"loggers": {"console": {"config": {"log_level": "DEBUG"}}}}
    job.execute_in_process(run_config=run_config)
    
  ```
  
  The output
  ```
ERROR LOG MSG
WARNING LOG MSG
2025-03-12 11:55:43 +0100 - dagster - DEBUG - job - 7f6576bf-caa7-4e1a-8cf3-00b3d9f177b0 - 25877 - RUN_START - Started execution of run for "job".
2025-03-12 11:55:43 +0100 - dagster - DEBUG - job - 7f6576bf-caa7-4e1a-8cf3-00b3d9f177b0 - 25877 - ENGINE_EVENT - Executing steps in process (pid: 25877)
2025-03-12 11:55:43 +0100 - dagster - DEBUG - job - 7f6576bf-caa7-4e1a-8cf3-00b3d9f177b0 - 25877 - LOGS_CAPTURED - Started capturing logs in process (pid: 25877).
2025-03-12 11:55:43 +0100 - dagster - DEBUG - job - 7f6576bf-caa7-4e1a-8cf3-00b3d9f177b0 - 25877 - ENGINE_EVENT - Finished steps in process (pid: 25877) in 0.89ms
2025-03-12 11:55:43 +0100 - dagster - DEBUG - job - 7f6576bf-caa7-4e1a-8cf3-00b3d9f177b0 - 25877 - RUN_SUCCESS - Finished execution of run for "job".

```

I also added a `config.yml` file in the same (root) dir.
```
loggers:
  console:
    config:
      log_level: DEBUG
```  

That did also not work.
Any clues what to change?

**Reply by SbstnErhrdt** [reply]
I got a reply in the slack channel by zach. I post it here for clarification purpose:

> Generally speaking you shouldn't really need to be logging within `@job` definitions... everything that is defined within `@job` is only executed when your code is loaded, not when the job executes. The `@job` definitions should really only call ops or graphs, Dagster won't capture the logs within `@job` because the context isn't constructed when job definitions are executed. The DEBUG log level you're setting in the run config only gets applied to loggers that are within `@op` definitions, and the `logging.basicConfig(level="DEBUG")` doesn't apply to Dagster loggers, I think because Dagster loggers returned from `dg.get_dagster_logger()` are constructed with specific configuration that doesn't use the global config set by logging.basicConfig

**Comment by SbstnErhrdt** [comment]
```python
import logging
import dagster as dg


@dg.op()
def process(context):
    logger = context.log
    logger.info(f"OP: INFO LOG MSG") # LOG
    logger.error(f"OP: ERROR LOG MSG") # LOG
    logger.warning(f"OP: WARNING LOG MSG") # LOG
    logger.debug(f"OP: DEBUG LOG MSG") # LOG

@dg.job(
    config={"loggers": {"console": {"config": {"log_level": "DEBUG"}}}}
)
def job() -> None:
    logger = dg.get_dagster_logger()
    logger.info(f"JOB: INFO LOG MSG") # NO LOG
    logger.error(f"JOB: ERROR LOG MSG") # LOG
    logger.warning(f"JOB: WARNING LOG MSG") # LOG
    logger.debug(f"JOB: DEBUG LOG MSG") # NO LOG
    process()



if __name__ == "__main__":
    logging.basicConfig(level="DEBUG")
    run_config = {
        "loggers": {
            "console": {
                "config": {
                    "log_level": "DEBUG"
                }
            },
        }
    }
    job.execute_in_process(run_config=run_config)

```

Within the `@op` logging works with the levels.
Within `@job` only `error` and `warning` works.

**Comment by whillas-yabble** [comment]
trying to get this going for a sensor. no joy
